src/app/clusters/time-synchronization-server/time-synchronization-cluster.cpp

```diff
@@ -136,151 +88,329 @@ static CHIP_ERROR UpdateUTCTime(uint64_t UTCTimeInChipEpochUs)
+            {
+                ReturnErrorOnFailure(encod.Encode(tzStore.timeZone));
+            }
+            return CHIP_NO_ERROR;
+        });
+        return err;
+    }
+    case DSTOffset::Id: {
+        CHIP_ERROR err = encoder.EncodeList([this](const auto & encod) -> CHIP_ERROR {
+            const auto & dstList = GetDSTOffset();
+            for (const auto & dstOffset : dstList)
+            {
+                ReturnErrorOnFailure(encod.Encode(dstOffset));
+            }
+
+            return CHIP_NO_ERROR;
+        });
+        return err;
+    }
+    case TimeZoneListMaxSize::Id:
+        return encoder.Encode<uint8_t>(CHIP_CONFIG_TIME_ZONE_LIST_MAX_SIZE);
+    case DSTOffsetListMaxSize::Id:
+        return encoder.Encode<uint8_t>(CHIP_CONFIG_DST_OFFSET_LIST_MAX_SIZE);
+    case LocalTime::Id: {
+        DataModel::Nullable<uint64_t> localTime;
+        CHIP_ERROR err = GetLocalTime(localTime);
+        if (err == CHIP_ERROR_INVALID_TIME)
+        {
+            return encoder.EncodeNull();
+        }
+        ReturnErrorOnFailure(err);
+        return encoder.Encode(localTime);
+    }
+    case ClusterRevision::Id:
+        return encoder.Encode(TimeSynchronization::kRevision);
+    case FeatureMap::Id:
+        return encoder.Encode(mFeatures);
+    case SupportsDNSResolve::Id:
+        return encoder.Encode(mSupportsDNSResolve);
+    case TimeZoneDatabase::Id:
+        return encoder.Encode(mTimeZoneDatabase);
+    case TimeSource::Id:
+        return encoder.Encode(mTimeSource);
+    case NTPServerAvailable::Id:
+        return encoder.Encode(mNTPServerAvailable);
+    default:
+        return Protocols::InteractionModel::Status::UnsupportedAttribute;
+    }
+}
+
+CHIP_ERROR TimeSynchronizationCluster::Startup(ServerClusterContext & context)
+{
+    ReturnErrorOnFailure(DefaultServerCluster::Startup(context));
+
+    mTimeSyncDataProvider.Init(context.storage);
+
+    Structs::TrustedTimeSourceStruct::Type tts;
+    if (mTimeSyncDataProvider.LoadTrustedTimeSource(tts) == CHIP_NO_ERROR)
+    {
+        mTrustedTimeSource.SetNonNull(tts);
+    }
+    if (LoadTimeZone() != CHIP_NO_ERROR)
+    {
+        ClearTimeZone();
+    }
+    if (LoadDSTOffset() != CHIP_NO_ERROR)
+    {
+        ClearDSTOffset();
+    }
+
+    // Set the granularity to none for now - this will force us to go to the delegate so it can
+    // properly report the time source
+    mGranularity = GranularityEnum::kNoTimeGranularity;
+
+    // This can error, but it's not clear what should happen in this case. For now, just ignore it because we still
+    // want time sync even if we can't register the delegate here.
+    CHIP_ERROR err = Server::GetInstance().GetFabricTable().AddFabricDelegate(this);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "Unable to register Fabric table delegate for time sync");
+    }
+    PlatformMgr().AddEventHandler(OnPlatformEventWrapper, reinterpret_cast<intptr_t>(this));
```

Comment:
This also returns a CHIP_ERROR. Maybe we should use some log_on_failure helper for these things, since realistically even if we fail there is not much we can do to recover.